File: apps/web/src/features/audit/api/list-security-events.ts
// apps/web/src/features/audit/api/list-security-events.ts

import { keepPreviousData, queryOptions, useQuery } from "@tanstack/react-query"

import type { SecurityEventsListResponse } from "@/features/audit/types"
import { apiRequest } from "@/lib/api/client"

export type ListSecurityEventsParams = {
  endpoint?: string | undefined
  eventType?: string | undefined
  ipAddress?: string | undefined
  limit?: number | undefined
  offset?: number | undefined
  occurredAfter?: string | undefined
  occurredBefore?: string | undefined
  userEmail?: string | undefined
}

export const securityEventsQueryKeys = {
  all: ["security-events"] as const,
  details: () => [...securityEventsQueryKeys.all, "detail"] as const,
  detail: (eventId: string) => [...securityEventsQueryKeys.details(), eventId] as const,
  lists: () => [...securityEventsQueryKeys.all, "list"] as const,
  list: (params: ListSecurityEventsParams = {}) =>
    [...securityEventsQueryKeys.lists(), params] as const,
}

async function listSecurityEvents({
  endpoint,
  eventType,
  ipAddress,
  limit = 50,
  offset = 0,
  occurredAfter,
  occurredBefore,
  userEmail,
}: ListSecurityEventsParams = {}) {
  return apiRequest<SecurityEventsListResponse>("/security-events/", {
    query: {
      endpoint,
      event_type: eventType,
      ip_address: ipAddress,
      limit,
      offset,
      occurred_after: toApiDateTime(occurredAfter),
      occurred_before: toApiDateTime(occurredBefore),
      user_email: userEmail,
    },
  })
}

function securityEventsQueryOptions(
  params: ListSecurityEventsParams = {},
  options: { enabled: boolean }
) {
  return queryOptions({
    queryKey: securityEventsQueryKeys.list(params),
    queryFn: () => listSecurityEvents(params),
    enabled: options.enabled,
    placeholderData: keepPreviousData,
    staleTime: 15_000,
  })
}

export function useSecurityEventsQuery(
  params: ListSecurityEventsParams = {},
  options: { enabled: boolean }
) {
  return useQuery(securityEventsQueryOptions(params, options))
}

function toApiDateTime(value: string | undefined) {
  return value ? new Date(value).toISOString() : undefined
}

File: apps/api/routes/security_events/list_security_events.py
# apps/api/routes/security_events/list_security_events.py

"""Route for listing security events."""

from datetime import datetime
from typing import Annotated

from fastapi import APIRouter, Query

from core.dependencies import AsyncDbSessionDep
from services.security import list_security_events_for_super_admin
from services.security.schemas import SecurityEventsListResponse

router = APIRouter()


@router.get("/")
async def list_security_events(
    db: AsyncDbSessionDep,
    limit: Annotated[int, Query(ge=1, le=200)] = 50,
    offset: Annotated[int, Query(ge=0)] = 0,
    event_type: Annotated[str | None, Query(max_length=100)] = None,
    user_email: Annotated[str | None, Query(max_length=320)] = None,
    ip_address: Annotated[str | None, Query(max_length=64)] = None,
    endpoint: Annotated[str | None, Query(max_length=500)] = None,
    occurred_after: Annotated[datetime | None, Query()] = None,
    occurred_before: Annotated[datetime | None, Query()] = None,
) -> SecurityEventsListResponse:
    return await list_security_events_for_super_admin(
        db,
        limit=limit,
        offset=offset,
        event_type=event_type,
        user_email=user_email,
        ip_address=ip_address,
        endpoint=endpoint,
        occurred_after=occurred_after,
        occurred_before=occurred_before,
    )
